mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala

```diff
@@ -0,0 +1,549 @@
+          mdef.symbol.annotations.map(_.original)
+        case anns => anns
+      }
+    }
+
+    private def typePos(tpe: cp.TypeTree) = {
+      tpe.original match {
+        case cp.AppliedTypeTree(tpt, _) => tpt.pos
+        case sel: cp.Select => sel.namePos
+        case _ => tpe.pos
+      }
+    }
+    // We need to collect named params since they will not show on fully typed tree
+    lazy val namedArgCache = {
+      val parsedTree = cp.parseTree(source)
+      parsedTree.collect { case arg @ cp.AssignOrNamedArg(_, rhs) =>
+        rhs.pos.start -> arg
+      }.toMap
+    }
+  }
+
+  def pickFromTraversed(tk:scala.meta.tokens.Token):List[cp.Tree] ={
+      val buffer = ListBuffer.empty[cp.Tree]
+        
+      for (node <- nodes){
+        if( node.pos.start ==tk.pos.start &&
+           node.pos.end ==tk.pos.end 
+        ) buffer.addAll(List(node.tree))
+      }
+
+      buffer.toList
+  }
+
+
+  /**
+    * returns (SemanticTokenType, SemanticTokenModifier) of @param tk
+    */
+  private def getTypeAndMod(tk:scala.meta.tokens.Token):(Int, Int,String) ={
+
+    // whether token is identifier or not
+    tk match {
+      case _: Token.Ident => // continue this method
+      case _ =>
+         //Non-Ident has no modifier.
+         return (typeOfNonIdentToken(tk), 0, strSep + "Non-Ident")
+    }
+
+    var logString = ""
+    logString += linSep + linSep  + "  Start:Ident Part getSemanticTypeAndMod"
+    logString += linSep + "  " + tk.name
+
+    val nodeList = pickFromTraversed(tk)
+    val node = if (nodeList.size ==0) null else nodeList(0)
+    if( node == null) return (-1,0,strSep + "Node-Nothing") // break
+
+    logString += linSep + "  ** Got node"
+    //get type
+    val sym = node.symbol
+    val typ =  if (sym.isValueParameter ) getTid(SemanticTokenTypes.Parameter)
+      else node.symbol.keyString match {
+          case kind.kDef => 
+              if (sym.isGetter || sym.isSetter ) getTid(SemanticTokenTypes.Variable)
+              else getTid(SemanticTokenTypes.Method)
+          case kind.kVal => getTid(SemanticTokenTypes.Variable)
+          case kind.kVar => getTid(SemanticTokenTypes.Variable)
+          case kind.KClass => getTid(SemanticTokenTypes.Class)
+          case kind.kType => getTid(SemanticTokenTypes.Type)
+          case kind.kTrait => getTid(SemanticTokenTypes.Interface)
+          case kind.kObject =>  getTid(SemanticTokenTypes.Class) // treat object as class
+          case kind.kPackage => getTid(SemanticTokenTypes.Namespace)
+          case _ => -1
+      }
+
+    //get moodifier
+    var mod:Int = 0
+    def addPwrToMod(place:Int)={
+      if (place != -1) mod += scala.math.pow(2, place).toInt
+    }
+
+    if (node.symbol.isAbstract) addPwrToMod(getMid(SemanticTokenModifiers.Abstract))
+    if (node.symbol.keyString==kind.kVal) addPwrToMod(
+          getMid(SemanticTokenModifiers.Readonly))
+
+    //return
+    return (typ,mod,logString)
```

Comment:
```suggestion
    (typ,mod,logString)
```
return is optional in Scala

Reply:
I'll do it.